### PR TITLE
database: Don't allow empty code host config

### PR DESCRIPTION
**Pull request by ryanslade**
We already have a db constraint to stop this in most cases, but when we
have encryption enabled then even an empty config is converted into an
encrypted value so we need to do the check in our application layer too.

While investigating this we already had code to validate our config but
it had not been included in the Upsert path. That is now fixed and tests
were updated where necessary.

Closes https://github.com/sourcegraph/sourcegraph/issues/41924

## Test plan

New unit tests added and older ones updated.

### CODE REVIEW

**Review by mollylogue — APPROVED** [review]
👍

**Review by unknwon — APPROVED** [review]
Nice catch!

**Comment by eseliger** [comment]
Should we instead say the config has to be _parseable_?

**Comment by ryanslade** [comment]
> Should we instead say the config has to be _parseable_?

@eseliger I'm going to look into this today. 

We already have the [MakeValidateExternalServiceConfigFunc](https://github.com/sourcegraph/sourcegraph/blob/fad6910997c7b627f7271c26461523e1a41212e3/internal/database/external_services.go#L335) function which appears to do what we want and even checks the JSON against our schema. I'll look into whether we've missed calling that somewhere or if there's a bug somewhere in that code.

**Comment by eseliger** [comment]
Might be that this PR would fix it, too (didn't dive too deep): https://github.com/sourcegraph/sourcegraph/pull/42039

**Review by sashaostrikov — APPROVED** [review]

**Comment by ryanslade** [comment]
> Might be that this PR would fix it, too (didn't dive too deep): #42039

That PR assumes we were doing the validation correctly in all cases, but
we weren't doing it for the Upsert case.

**Comment by ryanslade** [comment]
PTAL

File: internal/database/external_services.go [hunk 1]
```diff
@@ -661,7 +673,7 @@ func (e *externalServiceStore) Upsert(ctx context.Context, svcs ...*types.Extern
 			s.Config.Set(rawConfig)
 		}
 
-		if err := e.recalculateFields(s, rawConfig); err != nil {
+		if err := e.recalculateFields(s, string(normalized)); err != nil {
```
Comment (eseliger) [inline internal/database/external_services.go:676]: what is `normalized` vs `rawConfig`? Does that still contain the original input with all the comments, indentation etc?

**Review by eseliger — APPROVED** [review]

File: internal/database/external_services.go (hunk 1 above)
Comment (ryanslade) [reply internal/database/external_services.go:676]: Normalised it the JSON representation with comments removed, redacted fields populated, disallowed fields removed etc

**Comment by ryanslade** [comment]
This change has broken *a lot* of tests outside of the `internal/database` package. I'll go though and update them. I think in most cases it's because we use dummy code host config that doesn't conform to our schema.

File: enterprise/cmd/frontend/internal/batches/webhooks/gitlab_test.go [hunk 2]
```diff
@@ -90,6 +90,9 @@ func testGitLabWebhook(db *sql.DB) func(*testing.T) {
 			})
 
 			t.Run("malformed external service", func(t *testing.T) {
+				// TODO: Check with batch changes team
```
Comment (ryanslade) [inline enterprise/cmd/frontend/internal/batches/webhooks/gitlab_test.go:93]: @eseliger What do you think about this and the other one lower down

File: enterprise/cmd/frontend/internal/batches/webhooks/gitlab_test.go [hunk 3]
```diff
@@ -289,6 +292,9 @@ func testGitLabWebhook(db *sql.DB) func(*testing.T) {
 			})
 
 			t.Run("error from handleEvent", func(t *testing.T) {
+				// TODO: Check with batch changes team
```
Comment (ryanslade) [inline enterprise/cmd/frontend/internal/batches/webhooks/gitlab_test.go:295]: @eseliger What do you think about this?

File: enterprise/cmd/frontend/internal/batches/webhooks/bitbucketserver_test.go [hunk 4]
```diff
@@ -67,12 +67,13 @@ func testBitbucketServerWebhook(db database.DB, userID int32) func(*testing.T) {
...
 					Secret: secret,
 				},
+				Token: "abc",
 			})),
 		}
 
 		err := esStore.Upsert(ctx, extSvc)
 		if err != nil {
-			t.Fatal(t)
+			t.Fatal(err)
```
Comment (eseliger) [inline enterprise/cmd/frontend/internal/batches/webhooks/bitbucketserver_test.go:76]: good catch! wtf

File: enterprise/cmd/frontend/internal/batches/webhooks/gitlab_test.go (hunk 2 above)
Comment (eseliger) [reply enterprise/cmd/frontend/internal/batches/webhooks/gitlab_test.go:93]: yes, sounds like it!

File: enterprise/cmd/frontend/internal/batches/webhooks/gitlab_test.go (hunk 3 above)
Comment (eseliger) [reply enterprise/cmd/frontend/internal/batches/webhooks/gitlab_test.go:295]: can we make handleEvent fail otherwise?

**Review by eseliger — APPROVED** [review]

**Pull request merged**